### DISCUSSION TITLE
Feedback and ideas

### DISCUSSION THREAD

**Discussion by VykosX** [VykosX/ControlFlowUtils#1] [Announcements]
Feel free to leave any feedback and ideas! This is still very much a WIP.

**Comment by christian-byrne** [comment]
Is the Data Monitor node supporting formulas when the parameters are tensors? Things like `~%aux1% ^ %aux2% > %aux3` where aux1 and aux2 are masks and aux3 is an image. The AST parser seems to be converting tensors to strings? Can you explain how it works?

**Reply by christian-byrne** [reply]
I found the examples which seem to answer the question:

https://github.com/VykosX/ControlFlowUtils/blob/1f5708ac310ed1e325b08ee07ef22ad8bbc709b4/Helpers.py#L689-L692

But when I try this:

![Selection_380](https://github.com/user-attachments/assets/aaab8cd0-31f6-4871-901c-5af1db0b9710)

I get error:

<details>

<summary>Expand error traceback</summary>

```
>> DATA MONITOR [ 14 ]
TEXT:  %aux% * 2
PASSTHROUGH:  <class 'NoneType'> None
AUX INPUTS:  [tensor([[[[0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          ...,
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000]],

         [[0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          ...,
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000]],

         [[0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          ...,
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000]],

         ...,

         [[0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          ...,
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039]],

         [[0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          ...,
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039]],

         [[0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          ...,
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039]]]]), None, None, None, None]
PREV: 
OUTPUT TYPE:  FORMULA
REPLACING '%aux%' in '%aux% * 2' with 'tensor([[[[0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          ...,
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000]],

         [[0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          ...,
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000]],

         [[0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          ...,
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000]],

         ...,

         [[0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          ...,
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039]],

         [[0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          ...,
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039]],

         [[0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          ...,
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039]]]])'!
>> PROCESSED TEXT AFTER VARIABLE REPLACEMENT: 'tensor([[[[0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          ...,
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000]],

         [[0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          ...,
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000]],

         [[0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          ...,
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000],
          [0.3686, 0.5804, 1.0000]],

         ...,

         [[0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          ...,
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039]],

         [[0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          ...,
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039]],

         [[0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          ...,
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039],
          [0.3686, 1.0000, 0.4039]]]]) * 2'
<built-in method tensor of type object at 0x753a66b957c0>
[[[[[0.3686, 0.5804, 1.0], [0.3686, 0.5804, 1.0], [0.3686, 0.5804, 1.0], Ellipsis, [0.3686, 0.5804, 1.0], [0.3686, 0.5804, 1.0], [0.3686, 0.5804, 1.0]], [[0.3686, 0.5804, 1.0], [0.3686, 0.5804, 1.0], [0.3686, 0.5804, 1.0], Ellipsis, [0.3686, 0.5804, 1.0], [0.3686, 0.5804, 1.0], [0.3686, 0.5804, 1.0]], [[0.3686, 0.5804, 1.0], [0.3686, 0.5804, 1.0], [0.3686, 0.5804, 1.0], Ellipsis, [0.3686, 0.5804, 1.0], [0.3686, 0.5804, 1.0], [0.3686, 0.5804, 1.0]], Ellipsis, [[0.3686, 1.0, 0.4039], [0.3686, 1.0, 0.4039], [0.3686, 1.0, 0.4039], Ellipsis, [0.3686, 1.0, 0.4039], [0.3686, 1.0, 0.4039], [0.3686, 1.0, 0.4039]], [[0.3686, 1.0, 0.4039], [0.3686, 1.0, 0.4039], [0.3686, 1.0, 0.4039], Ellipsis, [0.3686, 1.0, 0.4039], [0.3686, 1.0, 0.4039], [0.3686, 1.0, 0.4039]], [[0.3686, 1.0, 0.4039], [0.3686, 1.0, 0.4039], [0.3686, 1.0, 0.4039], Ellipsis, [0.3686, 1.0, 0.4039], [0.3686, 1.0, 0.4039], [0.3686, 1.0, 0.4039]]]]]
!!! Exception during processing !!! Could not infer dtype of ellipsis
Traceback (most recent call last):
  File "/home/c_byrne/projects/comfy-testing-environment/ComfyUI/execution.py", line 323, in execute
    output_data, output_ui, has_subgraph = get_output_data(obj, input_data_all, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb)
                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/c_byrne/projects/comfy-testing-environment/ComfyUI/execution.py", line 198, in get_output_data
    return_values = _map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/c_byrne/projects/comfy-testing-environment/ComfyUI/execution.py", line 169, in _map_node_over_list
    process_inputs(input_dict, i)
  File "/home/c_byrne/projects/comfy-testing-environment/ComfyUI/execution.py", line 158, in process_inputs
    results.append(getattr(obj, func)(**inputs))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/c_byrne/projects/comfy-testing-environment/ComfyUI/custom_nodes/ControlFlowUtils/ControlFlowUtils.py", line 1543, in data_monitor
    text = safe_eval(str(text))
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/c_byrne/projects/comfy-testing-environment/ComfyUI/custom_nodes/ControlFlowUtils/Helpers.py", line 502, in safe_eval
    return _eval(node, variables)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/c_byrne/projects/comfy-testing-environment/ComfyUI/custom_nodes/ControlFlowUtils/Helpers.py", line 442, in _eval
    result = _eval(stmt, local_vars)
             ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/c_byrne/projects/comfy-testing-environment/ComfyUI/custom_nodes/ControlFlowUtils/Helpers.py", line 439, in _eval
    return _eval(node.value, local_vars)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/c_byrne/projects/comfy-testing-environment/ComfyUI/custom_nodes/ControlFlowUtils/Helpers.py", line 352, in _eval
    left = _eval(node.left, local_vars)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/c_byrne/projects/comfy-testing-environment/ComfyUI/custom_nodes/ControlFlowUtils/Helpers.py", line 417, in _eval
    return func(*args)
           ^^^^^^^^^^^
RuntimeError: Could not infer dtype of ellipsis
```

</detail>

**Reply by VykosX** [reply]
The Data Monitor is a multi-purpose node, but primarily it is meant for data visualization and manipulation.

You can use it as a multiline textbox that you can pick the output format of; you can hook something into passthrough so you can visualize the data; you can convert the data between different formats by selecting a different output format than the passthrough; you can manipulate the data using the aux inputs in conjunction with the Formula mode by using python syntax e.g. accessing list elements, dictionary values, etc; and finally you can run a limited subset of python statements that are deemed 'safe'. Rather than using `eval()` or `exec()` the data is parsed using the `safe_eval()` function in Helpers.py, which attempts to sanitize the data and ensure only a subset of operations are processed, those are defined that file and i've been slowly adding to it as I think of more things that would be useful to have but won't result in arbitrary code execution.

As for your original question. Currently the Data Monitor does not work correctly with Tensors when using Formula mode. I would love for it to, but I couldn't get it to work correctly. safe_eval is completely lost on how to handle them, they are understood as a string representation and it completely fails to parse the ellipses. I suspect when you try to print tensors or operate on them literally you get a very basic repr() of the data which is not a proper full serialization and it throws off AST entirely.

In short, if you're willing I would love a PR adding the ability to manipulate the tensors directly, because I've tried adding support for them via safe_eval for a few hours and just couldn't get it to work. I looked at your Github and it seems you've done something similar with a node that runs arbitrary Python code? If you'd like to add support for manipulating tensors directly in a similar way to how you're handling it, I'd be happy to merge that PR.

Thanks a lot for your interest and your contributions!

**Reply by VykosX** [reply]
> I found the examples which seem to answer the question:
> 
> https://github.com/VykosX/ControlFlowUtils/blob/1f5708ac310ed1e325b08ee07ef22ad8bbc709b4/Helpers.py#L689-L692
> 
> But when I try this:
> 
> ![Selection_380](https://private-user-images.githubusercontent.com/72887196/368953626-aaab8cd0-31f6-4871-901c-5af1db0b9710.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY3NDMzOTMsIm5iZiI6MTcyNjc0MzA5MywicGF0aCI6Ii83Mjg4NzE5Ni8zNjg5NTM2MjYtYWFhYjhjZDAtMzFmNi00ODcxLTkwMWMtNWFmMWRiMGI5NzEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE5VDEwNTEzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY5NGUxMzc5ODJiNzEyN2E0MjI5ZjRjMzM3ODViYjk2NjBjODAwY2QwMmRjMTcwY2EzZTk1OWFkZTJjOGMyZTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1d8oYuMBVxxmMbP0dzKobVQkpA9e4zP13JU6kNyaeQI)
> 
> I get error:
> Expand error traceback

What you're seeing was my attempt at trying to coerce gpt-4o into making tensors work with AST but i spent a good couple hours throwing things at a wall unsuccessfully. I even tried to have it define an exception for tensors and just use regular eval() on them. I don't think they can be handled like that, we probably need an entire separate function for parsing and handling tensor operations.

**Reply by christian-byrne** [reply]
So the issue is not with the AST parser but with the `DataMonitor.replace_vars` casting all types to strings? It seems there is a method to reference objects from outside the eval scope, but it's not working with tensors? Or am I doing something wrong?

**Reply by christian-byrne** [reply]
Sorry, I wrote many comments before refreshing and seeing your replies.

I can definitely PR something. Can you explain briefly the `__MEM__STORAGE__KEYS__` system to help me?

**Reply by VykosX** [reply]
> So the issue is not with the AST parser but with the `DataMonitor.replace_vars` casting all types to strings? It seems there is a method to reference objects from outside the eval scope, but it's not working with tensors? Or am I doing something wrong?

That's a really good point actually, I think you found the general issue, the typecasting that happens in `replace_vars`. That was added long before I thought to try and support Tensors and non-literal types and by the time the data reaches safe_eval it has already been converted to a string. Which means we'd need to add an exception for Tensors inside replace_vars for the passthroughs and aux inputs and then perhaps safe_eval might be able to handle them? I realize I never actually tried because I was working off a forced typecast I missed. Nothing like a fresh set of eyes scanning your code to find bugs you missed. You seem to be on the right track here friend, so if you wanna get at the bottom of this, by all means, I'm really enjoying this collaboration!

**Reply by VykosX** [reply]
> Sorry, I wrote many comments before refreshing and seeing your replies.
> 
> I can definitely PR something. Can you explain briefly the `__MEM__STORAGE__KEYS__` system to help me?

the `MEM__STORAGE__X` commands are a way to manipulate Memory Storages from within the Data Monitor node directly. Rather than using actual Memory Storage nodes which are limited only to a specific variable at a time, using those requests lets you work with the entire Memory Storage dictionary natively, allowing you to clear, set, get and access the names of all stored variables natively. There is a sample workflow included which makes use of Memory Storages and the Data Monitor manipulation requests in the form of `VykosX-Random Prompt Generator + Bifurcated Workflow with Universal Switch + Memory Storages Showcase.png` in the workflows folder so you can see it in action.

**Reply by VykosX** [reply]
> Sorry, I wrote many comments before refreshing and seeing your replies.
> 
> I can definitely PR something. Can you explain briefly the `__MEM__STORAGE__KEYS__` system to help me?

Btw if you'd like to make this collaboration a little easier and/or get ahold of me for more questions outside of the feedback posts, feel free to send a message over at Discord, my username is the same as on github (`vykosx`). I'm also on the official ComfyUI and L2 discord servers.  =)

Looking forward to your PR, it's wonderful seeing interest in the pack, I think there's still a lot that can be done in the pack to support all kinds of dynamic workflow systems, but it's just been me trying to figure things out so far.

**Reply by christian-byrne** [reply]
Thanks for all the info.

The string casting seems necessary based on what `ast.parse` expects. In `_eval` there is already logic for including local vars in eval scope. The solution seems to be to add a `isinstance(aux, torch.Tensor)` in `replace_vars`, and if the aux represents a tensor, just pass the reference to `safe_eval` like `safe_eval(expr, variables={aux: tensor_reference})`, instead of doing direct interpolation with `repr`. WDYT?

Not sure if there's a way to do it while maintaining the "safe" aspect entirely.

**Reply by VykosX** [reply]
That seems like a good solution to me. It's kind of what I attempted directly inside `safe_eval` initially but I missed the forced typecasting in `replace_vars` and just assumed the tensors had an implicit string conversion. I think perhaps check inside `replace_vars()` for the `isinstance` of Tensor and skip processing passthrough and auxs when it is in the same way `None` is being skipped?

Then you can call `safe_eval` just as you suggested for the `FORMULA` case, and assuming tensors are being properly handled on that it should work. I'm unsure if `safe_eval` can natively support tensor operations as is or if AST will throw a fit, but as long as the operators are being overloaded correctly and the tensors are modified in-place rather than resulting in copies I think it should work.

I'm also assuming tensors are inherently "safe" (i mean they're called .safetensors they're supposed to be free of pickles) so it's probably okay to allow them through directly. Hell I was willing to let them through to regular `eval()` as an explicit exception but even that didnt fix things due to that pesky string conversion hehe. This actually explains so much I feel silly not having noticed it sooner.

**Reply by christian-byrne** [reply]
Good point. Unsafe image tensors are not a realistic concern because uploading them isn't really possible and model tensors are serialized safely.

**Reply by VykosX** [reply]
> Good point. Unsafe image tensors are not a realistic concern because uploading them isn't really possible and model tensors are serialized safely.

Heya. How goes your PR to add tensor manipulation functionality? In the meantime I did some dirty patches in Data Monitor to just remove the ellipses from the tensor reprs so at least it doesn't crash safe_eval so that I could reenable being able to reference the inputs of Loop nodes within their conditions. I didn't want to mess with tensors until you've had your chance to get that added, so it's just a small addition.

**Comment by noahaust2** [comment]
Would it be possible to get some example workflows, especially for the cycle nodes? I think these nodes would be a lifesaver for a project I'm working on but I can't figure out the documentation

**Reply by VykosX** [reply]
There are many workflows in the Workflow folder included with the nodes, several of which showcase the usage of cycles.
